crawler/urlJobProcessor.js

```diff
@@ -1,36 +1,106 @@
 import { default as fetch } from 'node-fetch';
+import { default as cheerio } from 'cheerio';
+import { default as urlModule } from 'url';
 
 const probes = [];
 
-function configure(/* opts */) {
-  // TODO: Set up probes. They'll be specified somehow in opts and we
-  // should be able to load each probe individually. Maybe something
-  // like `require('./probe-' + probeName + '.js');`
+function configure(opts) {
+  if (!opts.probes || !opts.probes.length) {
+    return Promise.reject(new Error('At least one urlJobProcessor probe must be specified'));
+  }
+
+  const ret = [Promise.resolve()];
+
+  opts.probes.forEach((probeOpts) => {
+    if (!probeOpts.name) {
+      return Promise.reject(new Error('Probe without name encountered'));
+    }
+
+    // We allow the probe to be specified by name or by path. If specified by name,
+    // we expect it to be a known subdir.
+    const path = probeOpts.path || `./urlJobProcessorProbes/${probeOpts.name}.js`;
+    const probeModule = require(path).default;
+
+    if (!probeModule.runProbe) {
+      return Promise.reject(new Error(`Probe ${probeOpts.name} does not implement runProbe`));
+    }
+
+    let p = Promise.resolve();
+    if (probeModule.configure) {
+      p = probeModule.configure(probeOpts.opts || {});
+    }
+
+    ret.push(p.then(() => {
+      probes.push({
+        name: probeOpts.name,
+        probeFn: probeModule.runProbe,
+      });
+    }));
+  });
+
+  return Promise.all(ret);
 }
 
-function fetchAllResources(url) {
-  const ret = { html: '', js: '', sw: '' };
-  return fetch(`http://${url}`).then((response) => {
-    return response.text().then((htmlText) => {
-      ret.html += htmlText;
+function fetchAllResources(urlArg) {
+  const url = urlModule.parse(`http://${urlArg}`);
+
+  // TODO: Pass user agent in headers
+  return fetch(`${url.href}`, { headers: '' }).then((mainResponse) => {
+    // We pass along the response object that was returned after
+    // our fetch of the main page. As an optimization, we don't
+    // clone the response here. Instead, we pass the body text
+    // in a separate arg.
+    const responses = [mainResponse];
+    return mainResponse.text().then((htmlText) => {
+      const $ = cheerio.load(htmlText);
+      const scripts = [];
+      const promises = [];
+
+      $('script').each((index, elem) => {
```

Comment:
Fetching the scripts for each site, even when not on HTTPS adds a lot of overhead (think of the 1Mb+ JavaScript served these days). Any way that we can delay this until we actually hit a HTTPS site would be great.